librz/bp/serialize_bp.c

```diff
@@ -0,0 +1,146 @@
+// SPDX-FileCopyrightText: 2021 DMaroo <dhruvsmaroo@gmail.com>
+// SPDX-License-Identifier: LPGL-3.0-only
+
+#include <rz_util/rz_serialize.h>
+#include <rz_bp.h>
+
+RZ_API void rz_serialize_bp_save(RZ_NONNULL Sdb *db, RZ_NONNULL RzBreakpoint *bp) {
+	rz_return_if_fail(db && bp);
+
+	RzListIter *iter;
+	RzBreakpointItem *bp_item;
+	rz_list_foreach (bp->bps, iter, bp_item) {
+		PJ *j = pj_new();
+		if (!j) {
+			return;
+		}
+		pj_o(j);
+		pj_kn(j, "addr", bp_item->addr);
+		pj_ks(j, "bbytes", (char *)bp_item->bbytes);
+		pj_ks(j, "cond", bp_item->cond);
+		pj_ks(j, "data", bp_item->data);
+		pj_kn(j, "delta", bp_item->delta);
+		pj_ki(j, "enabled", bp_item->enabled);
+		pj_ks(j, "expr", bp_item->expr);
+		pj_ki(j, "hits", bp_item->hits);
+		pj_ki(j, "hw", bp_item->hw);
+		pj_ki(j, "internal", bp_item->internal);
+		pj_kN(j, "module_delta", bp_item->module_delta);
+		pj_ks(j, "module_name", bp_item->module_name);
+		pj_ks(j, "obytes", (char *)bp_item->obytes);
+		pj_ki(j, "perm", bp_item->perm);
+
+		pj_ka(j, "pids");
+		for (int i = 0; i < RZ_BP_MAXPIDS; i++) {
+			pj_i(j, bp_item->pids[i]);
+		}
+		pj_end(j);
+
+		pj_ki(j, "size", bp_item->size);
+		pj_kb(j, "swstep", bp_item->swstep);
+		pj_ki(j, "togglehits", bp_item->togglehits);
+		pj_ki(j, "trace", bp_item->trace);
+		pj_end(j);
+		sdb_set(db, bp_item->name, pj_string(j), 0);
+		pj_free(j);
+	}
+}
+
+static bool bp_load_cb(void *user, const char *k, const char *v) {
+	RzList *list = user;
+	char *json_str = strdup(v);
+	if (!json_str) {
+		return true;
+	}
+
+	RzJson *json = rz_json_parse(json_str);
+	if (!json) {
+		free(json_str);
+		return true;
+	}
+	if (json->type != RZ_JSON_OBJECT) {
+		goto return_goto;
+	}
+
+	const RzJson *child;
+	RzBreakpointItem *bp_item = RZ_NEW0(RzBreakpointItem);
+	if (!bp_item) {
+		goto return_goto;
+	}
+	bp_item->name = strdup(k);
+
+	for (child = json->children.first; child; child = child->next) {
+		if (child->type == RZ_JSON_INTEGER) {
```

Comment:
I saw other code using `RZ_KEY_PARSER_JSON`, is that possible here as well?

Reply:
I'll have a look. Thanks.

Reply:
https://github.com/rizinorg/rizin/blob/dev/librz/analysis/serialize_analysis.c#L637-L664

Looks like we can also make a key parser to do this job, but I don't exactly know how helpful it will be. @thestr4ng3r Your thoughts, please.

Reply:
Yes, the key parser should be used here. It's both more concise and faster since it avoids the `strcmp` chaining.